Boa 0.15 release (#44)
boa-dev/boa-dev.github.io: posts/2022-06-05-boa-release-15.md

---
layout: post
tags: post
title: "Boa release v0.15"
author: Boa Developers
---

## Summary

Boa v0.15 is now available! After around 2.5 months of development, we are very happy to present you the newest release of the

> Boa JavaScript engine. Boa makes it easy to embed a JS engine in your projects, and you can even use it from webassembly. See

Reviewer: ```suggestion
Boa JavaScript engine. Boa makes it easy to embed a JS engine in your projects, and you can even use it from WebAssembly. See
```

in the official ECMAScript Test Suite (Test262). The engine now passes 56,372 tests, coming from 43,986 in Boa 0.14
(28.1% increase), and we have closed 18 issues and merged 58 pull requests. You can check the full list of changes
[here](https://github.com/boa-dev/boa/blob/v0.15/CHANGELOG.md), and the full information on conformance
[here](https://boa-dev.github.io/boa/test262/).

## New ECMAScript features

While there are only a few big new features in this release, there are a lot of fixes for existing features that should enable
many more JavaScript programs to execute correctly. For a detailed list checkout the
[changelog](https://github.com/boa-dev/boa/blob/v0.15/CHANGELOG.md).

### Support for Classes

With this new release boa ships with support for ECMAScript Classes. While not all features are implemented, most basic
functionality should work as expected.

```javascript
class Rectangle {
  constructor(height, width) {
    this.height = height;
    this.width = width;
  }

  calcArea() {
    return this.height * this.width;
  }

  get area() {
    return this.calcArea();
  }
}

let r = new Rectangle(2, 4);
r.calcArea(); // 8
```

### Support for `eval()` and `Function()`

> While especially `eval()` is one of the less popular JavScript features, many tests in the ECMAScript Test Suite use it. For a

Reviewer: ```suggestion
While `eval()` is one of the less popular Javascript features, many tests in the ECMAScript Test Suite use it. For a
```

better representation of passing tests and to fully comply with the specification we have implemented it and the similar
`Function()` constructor. We would like to echo the mdn docs and point out to
[Never use eval()!](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval#never_use_eval!)

```javascript
eval('console.log("Never use eval()!")');

let f = Function("arg", 'console.log("This is also a very " + arg + " idea!")');
f("bad");
```

## Regaining Performance

After moving from an AST based execution to our Virtual Machine we observed performance losses in some of our benchmarks. We
found out that the root-cause of these losses was a single `format!()` call in the hot-path of the VM. Our assumption was that
the formatting logic would be optimized out by rustc or llvm, because it was used as an argument to an empty (feature-gated)

> function. After replacing the `format!()` call with a `&'static str` the expected optimization took place and we where back to

Reviewer: ```suggestion
function. After replacing the `format!()` call with a `&'static str` the expected optimization took place and we were back to
```

previous performance levels. For more details check out [#1973](https://github.com/boa-dev/boa/pull/1973) for the changes and
some discussion about this issue. Thanks to @pdogr for finding the issue and contributing a fix.

## Starting work on Internationalization

> ECMAScript specifies an Internationalization API trough the `Intl` global object. We have started implementing many internal

Reviewer: ```suggestion
ECMAScript specifies an Internationalization API through the `Intl` global object. We have started implementing many internal
```

> functions and structures to provide the `Intl` object in boa. To take advantage of existing work in this space we have started
> integration the [ICU4X Project](https://github.com/unicode-org/icu4x) into boa. Thanks to @NorbertGarfield and @jedel1043 for

Reviewer: ```suggestion
functions and structures to provide the `Intl` object in boa. To take advantage of existing work in this space we have started
integrating the [ICU4X library](https://github.com/unicode-org/icu4x) into boa. Thanks to @NorbertGarfield and @jedel1043 for
```

their continued work on this topic.

## Rust Wrappers for Javascript Objects

We have started the implementation of wrappers around built-in Javascript objects like `Array`. These Rust types make it easy
to work with Javascript objects in Rust and provide a type safe abstraction around them. Many of these wrappers are not
implemented yet and may be a good place to start contributing to boa. If you are interested, check out the
[tracking issue](https://github.com/boa-dev/boa/issues/2098).

## How can you contribute to Boa?

In March, boa opened financial contributions in its [OpenCollective page](https://opencollective.com/boa). Since then, we have
to thank Demergent Labs, a company using Boa in the Internet Computer, and that has decided to sponsor us, and to Clemens

> Koza, who is our first backer! We are already thinking on how we will use these contributions to improve how we develop Boa.

Reviewer: ```suggestion
In March, boa opened financial contributions on its [OpenCollective page](https://opencollective.com/boa). Since then, we have
to thank Demergent Labs, a company using Boa in the Internet Computer that has decided to sponsor us, and Clemens
Koza, who is our first backer! We are already thinking on how we will use these contributions to improve how we develop Boa.
```
